00-programming-fundamentals/scientific-debugging.md

```diff
@@ -0,0 +1,348 @@
+# Scientific Debugging
+
+## Learning Goals
+
+By the end of this lesson, students should be able to...
+
+- List the steps of the scientific method
+- Apply the scientific method to the debugging process
+- Define the term _repro_
+- Evaluate experiments in terms of _cost_ and _risk_
+- Use _controlled experiments_ to methodically investigate bugs
+
+## Introduction: How do you Debug?
+
+Many programmers, especially early in their careers, take an ad-hoc approach to debugging. This works fine for simple bugs, but is likely to fall apart as programs become more and more complex. The goal of this lesson is to define a process to make debugging efficient, repeatable and scalable.
+
+**Think/pair/share** (5/5/10 minutes): Spend some time reflecting on your debugging process. When you discover a bug in your code, how do you proceed? What steps do you follow to fix it?
```

Comment:
I'd like to give the students one or two concrete examples of bugs that they've likely debugged, just to get their minds working...

Consider:

**Think/pair/share** (5/5/10 minutes): Imagine a bug you see when something gets printed to the console twice when you expected it to be printed once. Spend some time reflecting on your debugging process. When you discover a bug in your code, how do you proceed? What steps do you follow to fix it?

Or consider adding a concrete example to the paragraph above?